File: src/com/AYC/canalguide/canalparser/NavInfoMarker.java
package com.AYC.canalguide.canalparser;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import com.AYC.canalguide.R;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

/**
 * This class holds marker information for buoys, beacons and some other navigation information
 * 
 * @author James O'Leary
 *
 */
public class NavInfoMarker extends MapMarker implements Serializable {

	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	
	private static final BitmapDescriptor greenBuoyIcon = 
			BitmapDescriptorFactory.fromResource(R.drawable.mmi_green_buoy);
	
	private static final BitmapDescriptor redBuoyIcon = 
			BitmapDescriptorFactory.fromResource(R.drawable.mmi_red_buoy);
	
	private static final BitmapDescriptor greenBeaconIcon = 
			BitmapDescriptorFactory.fromResource(R.drawable.mmi_green_beacon);

	private static final BitmapDescriptor redBeaconIcon = 
			BitmapDescriptorFactory.fromResource(R.drawable.mmi_red_beacon);
	
	private static final BitmapDescriptor bridgeIcon = 
			BitmapDescriptorFactory.fromResource(R.drawable.mmi_bridge);
	
	private String shore;
	private String featureUrl;
	private String featureColor;
	private int channelWidth;
	private int southWestDepth;
	private int middleDepth;
	private String middleDepthUrl;
	private int northEastDepth;
	private String overheadClearance;
	private String noaaPage;
	private String noaaPageUrl;
	
	public NavInfoMarker(LatLng latLng, String feature, double mile, String shore, 
			String featureUrl, String featureColor, int channelWidth, int southWestDepth, 
			int middleDepth, String middleDepthUrl, int northEastDepth, String overheadClearance, 
			String noaaPage, String noaaPageUrl){
		super(latLng, feature, "", mile);
		this.shore = shore;
		this.featureUrl = featureUrl;
		this.featureColor = featureColor;
		this.channelWidth = channelWidth;
		this.southWestDepth = southWestDepth;
		this.middleDepth = middleDepth;
		this.middleDepthUrl = middleDepthUrl;
		this.northEastDepth = northEastDepth;
		this.overheadClearance = overheadClearance;
		this.noaaPage = noaaPage;
		this.noaaPageUrl = noaaPageUrl;		 
	}
	

	public String getShore(){
		return shore;
	}

	public String getFeatureUrl(){
		return featureUrl;
	}

	public String getFeatureColor(){
		return featureColor;
	}

	public int getChannelWidth(){
		return channelWidth;
	}

	public int getSouthWestDepth(){
		return southWestDepth;
	}

	public int getMiddleDepth(){
		return middleDepth;
	}
	
	public String getMiddleDepthUrl(){
		return middleDepthUrl;
	}
	
	public int getNorthEastDepth(){
		return northEastDepth;
	}
	
	public String getOverheadClearance(){
		return overheadClearance;
	}
	
	public String getNoaaPage(){
		return noaaPage;
	}

	public String getNoaaPageUrl(){
		return noaaPageUrl;
	}


	@Override
	public MarkerOptions getMarkerOptions() {
		BitmapDescriptor bd = getBitmapDescriptor();
		if(bd != null)
			return new MarkerOptions()
				.title(getTitle())
				.position(new LatLng(lat, lng))
				.snippet(getSnippet())
				.anchor(0.5f, 0.5f)
				.icon(bd);
		else
			return null;
	}
	
	@Override
	public String getTitle(){
		return name;
	}
	
	@Override
	public String getSnippet(){
		return "Mile " + mile + 
				(isNotBlank(southWestDepth) ? ", SW Depth=" + southWestDepth : "") + 
				(isNotBlank(middleDepth) ? ", Middle Depth=" + middleDepth : "") + 
				(isNotBlank(northEastDepth) ? ", NE Depth=" + northEastDepth : "") + 
				(isNotBlank(overheadClearance) ? ", Overhead Clearance=" + overheadClearance : "");
	}
	
	public BitmapDescriptor getBitmapDescriptor(){
		
		if(name.toLowerCase().contains("buoy")){
			if(featureColor.equalsIgnoreCase("green"))
				return greenBuoyIcon;	//BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_GREEN);
			else if(featureColor.equalsIgnoreCase("red"))
				return redBuoyIcon;	//BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_RED);
		}
		else if(name.toLowerCase().contains("beacon")){
			if(featureColor.equalsIgnoreCase("green"))
				return greenBeaconIcon;	//BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_GREEN);
			else if(featureColor.equalsIgnoreCase("red"))
				return redBeaconIcon;	//BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_RED);
		}
		else if(name.toLowerCase().contains("bridge"))
			return bridgeIcon;
		
		return null;
	}

	@Override
	public MapMarker cloneWithoutMarker(){
		return new NavInfoMarker(new LatLng(lat, lng), name, mile, shore, featureUrl, 
				featureColor, channelWidth, southWestDepth, middleDepth, middleDepthUrl, 
				northEastDepth, overheadClearance, noaaPage, noaaPageUrl);
	}
	
	public static List<MapMarker> readMarker(XmlPullParser parser)
			throws XmlPullParserException, IOException {
        
    	List<MapMarker> mapMarkers = new ArrayList<MapMarker>();
		
        double lat = 0, lng = 0;
        String name = null;	
     	double mile = 0;
    	String shore = null;
    	String featureUrl = null;
    	String featureColor = null;
    	int channelWidth = 0;
    	int southWestDepth = 0;
    	int middleDepth = 0;
    	String middleDepthUrl = null;
    	int northEastDepth = 0;
    	String overheadClearance = null;
    	String noaaPage = null;
    	String noaaPageUrl = null;

		 try{
		 String tag;
		 int event = parser.getEventType();
		 while(event != XmlPullParser.END_DOCUMENT){
			try{
				while(parser.getEventType() != XmlPullParser.END_TAG)
					parser.next();
				parser.nextTag();
			} catch(XmlPullParserException e){
				log("Returning " + mapMarkers.size() + " NavInfoMarkers from catch");
				return mapMarkers;
			}
		    tag = parser.getName();
		    if (tag.equals("channelinfo")) {
		    	lat = parseDouble(parser.getAttributeValue(null, "latitude"));
		    	lng = parseDouble(parser.getAttributeValue(null, "longitude"));
		    	name = parser.getAttributeValue(null, "feature");
		    	mile = Double.parseDouble(parser.getAttributeValue(null, "mile"));
		    	shore = parser.getAttributeValue(null, "shore");
		    	featureUrl = parser.getAttributeValue(null, "feature_url");
		    	featureColor = parser.getAttributeValue(null, "feature_color");
		    	channelWidth = parseInt(parser.getAttributeValue(null, "channel_width"));
		    	southWestDepth = parseInt(parser.getAttributeValue(null, "south_west_depth"));
		    	middleDepth = parseInt(parser.getAttributeValue(null, "middle_depth"));
		    	middleDepthUrl = parser.getAttributeValue(null, "middle_depth_url");
		    	northEastDepth = parseInt(parser.getAttributeValue(null, "north_east_depth"));
		    	overheadClearance = parser.getAttributeValue(null, "overhead_clearance");
		    	noaaPage = parser.getAttributeValue(null, "noaa_page");
		    	noaaPageUrl = parser.getAttributeValue(null, "noaa_page_url");
		    
		    	if(lat != -1 || lng != -1)
		    		mapMarkers.add(new NavInfoMarker(new LatLng(lat, lng), name, mile, shore, featureUrl, 
	        				featureColor, channelWidth, southWestDepth, middleDepth, middleDepthUrl, 
	        				northEastDepth, overheadClearance, noaaPage, noaaPageUrl));	
	            
	            event = parser.next();   
		    }
		 }
		 }catch(Exception e){
			 e.printStackTrace();
		 }
		 
		 log("Returning " + mapMarkers.size() + " NavInfoMarkers");
		 return mapMarkers;
	}
		
	public String toString(){
		return super.toString() + shore + " " + featureUrl + " " + featureColor + " " + 
				channelWidth + " " + southWestDepth + " " + middleDepth + " " + middleDepthUrl + 
				" " + northEastDepth + " " + overheadClearance + " " + noaaPage + " " + noaaPageUrl;
	}
	
	private static void log(String msg){
		log("NavInfoMarker", msg);
    }

}

File: app/src/main/java/com/ayc/canalguide/data/entities/NavInfoMarker.kt
package com.ayc.canalguide.data.entities

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.TypeConverters
import com.ayc.canalguide.R
import com.ayc.canalguide.data.Constants
import com.ayc.canalguide.data.NavInfoType
import com.ayc.canalguide.utils.converters.NavInfoTypeConverter
import com.google.android.gms.maps.model.BitmapDescriptor
import com.google.android.gms.maps.model.BitmapDescriptorFactory

/*
<channelinfo mile="145.00" shore="west" feature="marina: Springer's" feature_url="http://www.springersmarina.com" feature_color=""
channel_width="300" south_west_depth="" middle_depth="" middle_depth_url="" north_east_depth="" overhead_clearance=""
noaa_page="chart viewer" noaa_page_url="http://www.charts.noaa.gov/OnLineViewer/12348.shtml" latitude="42.637241" longitude="-73.752825"/>
 */
@Entity(tableName = "navinfo_marker")
@TypeConverters(NavInfoTypeConverter::class)
data class NavInfoMarker (
        @Ignore
        override val markerId: Int = 0,
        @Ignore
        override val lat: Double,
        @Ignore
        override val lng: Double,
        @Ignore
        override val name: String,
        @Ignore
        override val bodyOfWater: String?,
        @Ignore
        override val mile: String,

        val shore: String?,
        val featureUrl: String?,
        val featureColor: String?,
        val channelWidth: String?,
        val southWestDepth: String?,
        val middleDepth: String?,
        val middleDepthUrl: String?,
        val northEastDepth: String?,
        val overheadClearance: String?,
        val noaaPage: String?,
        val noaaPageUrl: String?,
        val navInfoType: NavInfoType,
        val apiId: Int
): MapMarker(lat, lng, name, bodyOfWater, mile, markerId) {


    @Ignore
    val waterFlowEastWest = Constants.navInfoRegionsEastWestWaterflow.contains(apiId)


    fun getDepthSubtext(): String? {
        var subText = ""
        if (!northEastDepth.isNullOrBlank()) subText += "${if (waterFlowEastWest) "North" else "East"} Depth: $northEastDepth\n"
        if (!southWestDepth.isNullOrBlank()) subText += "${if (waterFlowEastWest) "South" else "West"} Depth: $southWestDepth\n"

        // Remove extra new line character
        return if(subText.isNotBlank()) subText.substring(0, subText.length - 1) else null
    }

    override fun getSnippet() = "Mile " + mile +  //(isNotBlank(southWestDepth) ? ", SW Depth=" + southWestDepth : "") +
            (if (isNotBlank(middleDepth)) ", Middle Depth=$middleDepth" else "") +  //(isNotBlank(northEastDepth) ? ", NE Depth=" + northEastDepth : "") +
            if (isNotBlank(overheadClearance)) ", Overhead Clearance=$overheadClearance" else ""

    override fun getMarkerOptions() = super.getMarkerOptions()?.icon( getBitmapDescriptor() )

    //val bitmapDescriptor: BitmapDescriptor? by lazy {
    private fun getBitmapDescriptor(): BitmapDescriptor? = when (navInfoType) {
        NavInfoType.GreenBuoy -> greenBuoyIcon
        NavInfoType.GreenBeacon -> greenBeaconIcon
        NavInfoType.RedBuoy -> redBuoyIcon
        NavInfoType.RedBeacon -> redBeaconIcon
        NavInfoType.OtherBeacon -> otherBeaconIcon
        NavInfoType.Bridge -> bridgeIcon
        NavInfoType.Unknown -> null
    }

    companion object {
        private val greenBuoyIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_green_buoy)
        private val redBuoyIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_red_buoy)
        private val greenBeaconIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_green_beacon)
        private val redBeaconIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_red_beacon)
        private val otherBeaconIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_other_beacon)
        private val bridgeIcon = BitmapDescriptorFactory.fromResource(R.drawable.mmi_bridge)
    }

}